upload-api/BRIDGE.md

```diff
@@ -0,0 +1,48 @@
+# HTTP-UCAN Bridge
+
+## Summary
+
+We have implemented a "bridge" that allows w3up users to interact with the service
+without implementing the UCAN invocation wire protocols. 
+
+A user can submit an HTTP request like (simplified for clarity):
+
+```
+POST /bridge
+Authorization: MYmI4NWUwMGFhNzNlZDlkM2Y2NDYxYWEwZjk1NDdjOWY=
```

Comment:
wdyt of implementing this instead? https://github.com/ucan-wg/ucan-http-bearer-token

Reply:
If we use UCAN bearer token we have to know the inputs, create and sign a UCAN prior to the request. The point of the bridge is to allow a client that does not know how to do this to still use the service.

Reply:
Sure, this gateway may not technically match one requirement in that spec https://github.com/ucan-wg/ucan-http-bearer-token?tab=readme-ov-file#332-insufficient-capability-scope

But other than that enough is similar that it seems like we could reuse the syntax and a lot of other thinking in that spec, e.g.:
* using a syntactically valid HTTP Authorization header instead of the one here that lacks an `<auth-scheme>`
* a way of dealing with large UCAN chains that is different than what this PR does, but unlike this PR keeps the authorization orthogonal from the request body (in headers) and not tightly coupled to any one request body media type (in this case JSON).
* encodes the UCAN as its canonical JWT encoding instead of a multibased car file. I thought this feature was motivated by people not wanting to have to deal with new encoding formats compared to old.web3.storage, and the UCAN signatures are over the canonical JWT, so it seems ideal to avoid novel encoding formats here and use a JWT like old.web3.storage does. It's also how the UCAN spec says is the canonical way of encoding UCANs ('UCANs MUST be formatted as JWTs').
* I'm not that attached to anyone solution here. But partially I wanted to link to that spec bcause @Gozala's writeup you linked to below says "They POST to the RPC gateway endpoint passing UCAN delegation as JWT bearer token" so passing something other than a bearer token, and other than a JWT, felt like it was deviating from that plan as described.

If this goes down the path of not reusing that fairly well defined way of authorizing http requests using UCANs and invents a new one, please at least add an [`<auth-scheme>`](https://www.rfc-editor.org/rfc/rfc9110#name-credentials) to this Authorization header value (e.g. `Bearer`) even if the rest of the value is not a UCAN JWT.

Reply:
as an aside, we can put the secret seed into a UCAN as a UCAN fct like 'did:zAlice derives from this KDF and this seed'. I can't think of a reason we need to keep them separate?

Reply:
1) +1 to adding an <auth-scheme> - definitely need to do this
2) I'm hesitant to use the `ucan-http-bearer-token` spec because it seems to implicitly limit the complexity of UCANs that can be used to authorize a request - the spec even notes that "the limit of [the size of an Authorization header in] most browsers [is] 8KB." I'm open to an argument that 8KB is a high enough limit that we should just accept it, but it imho it feels bad to bake this limit into our bridge without a compelling benefit (and imho "compliance with a spec" is not particularly compelling on its own in this case).
3) As @alanshaw mentioned above, we're trying to avoid forcing users of this bridge to implement UCAN signing, and I think that's directly at odds with using the ucan-http-bearer-token spec - in other words, using that spec feels like a total non-starter given the goals of this project - am I misunderstanding?
4) @bengo what you're proposing is a pretty different direction than what I've proposed - I am suggesting we only put the "identity" into this header and you're proposing we try to get the full authorization chain into it. In your proposed solution, where do you imagine the "identity" being passed? HTTP Basic auth does this by :-separating the "username" and "password" - do you think that would be a good idea or do you have other ideas about how this could work?

Reply:
whoop - worth noting that you answered (4) above! 

One design goal that I haven't stated anywhere afaik is to keep the really sensitive secret token isolated to the Authorization header - I think there's less of a chance intermediaries (or our systems) will log the Authorization header somewhere than the body - put another way, I'd expect the body of this request to end up in something equivalent to our UCAN log, and that's much worse if it contains both the private key AND the delegation UCAN(s) than if it just contained the delegation. 

If we do decide to put both the secret AND the delegation in the Authorization field then I'm much less concerned, and what you're proposing could work, though (I think) it would mean that we'd mix up the UCAN invocation inputs and the secret key, and would need to be diligent to filter out the key before we put the inputs into the invocation that we create and execute...

Reply:
Oh - reading the `ucan-http-bearer-token` spec just now I realized that it doesn't really specify whether the UCAN in the bearer token should be an invocation or a delegation. I do think we could use that spec to pass delegations! We'd just need to figure out where we'd like the `secret` to live. I'm still hesitant to go down this path because of the 8kb limit, but that is definitely the least strong of my objections above!

@Gozala would definitely love your thoughts here!

Reply:
Thanks for clarifying @gobengo. I had assumed you meant to use it to send an invocation which we wouldn't be able to do, but agreed it makes sense to follow for sending the delegation. I like that it moves the auth data out of the invocation instruction in the body.

We could just use an `X-Auth-Secret` header. Cloudflare for example, uses [headers like this](https://developers.cloudflare.com/api/) in it's own API. I appreciate the comment about not wanting to include this header in logs but in practice `Authorization` is just as susceptible as any other header to being logged.

I'm not keen on putting the secret in the facts of the delegation, I think we _would_ end up logging this UCAN and so would be logging secrets. It also wouldn't allow alice to be delegated permissions from bob without bob knowing alice's secret.

I'm ok with an 8KB limit. If it gets problematic, maybe we send a multipart request with extra UCANs in the body (or something)...but we can implement that if/when we need it.

re: multiple invocations - I assume a single HTTP request translates into a single UCAN invocation to our service. If multiple requests are sent then I think it won't necessarily translate into a single request to our service. i.e. I assume the handler at the bridge would be called for every request on a keepalive connection, causing multiple requests to our service upstream.

@travis WDYT can we move forward with this?

Reply:
yea after noodling on this I'm also pro-`ucan-http-bearer-token`! agree re: possible mitigation strategies if the `Authorization` header gets too big. 

I'll get this updated, I think this is probably enough shared agreement to move forward with at least another iteration!